File: firebase_app_distribution_android/android/src/main/kotlin/dev/fluttercommunity/firebase_app_distribution/FirebaseAppDistributionPlugin.kt
package dev.fluttercommunity.firebase_app_distribution

import android.content.Context
import androidx.annotation.NonNull
import com.google.firebase.appdistribution.FirebaseAppDistribution

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class FirebaseAppDistributionPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var progressChannel: EventChannel
    private var context: Context? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "firebase_app_distribution_android")
        channel.setMethodCallHandler(this)
        progressChannel = EventChannel(flutterPluginBinding.binaryMessenger, "firebase_app_distribution_android/download_progress")
        progressChannel.setStreamHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        val firebaseAppDistribution = FirebaseAppDistribution.getInstance()
        when(call.method) {
            "updateIfNewReleaseAvailable" -> {
                firebaseAppDistribution.updateIfNewReleaseAvailable()
                    .addOnProgressListener { progress ->
                        emitProgress(progress.apkBytesDownloaded, progress.apkFileTotalBytes, progress.updateStatus.name)
                    }
                    .addOnSuccessListener {
                        result.success(null)
                    }
                    .addOnFailureListener {
                        result.error("UPDATE_FAILED", "Can not update app", "updateIfNewReleaseAvailable() failed with $it")
                    }
            }
            "checkForNewRelease" -> {
                firebaseAppDistribution.checkForNewRelease().addOnSuccessListener { release ->
                    if (release == null) {
                        result.success(null)
                    } else {
                        result.success(mapOf(
                            "displayVersion" to release.displayVersion,
                            "versionCode" to release.versionCode,
                            "releaseNotes" to release.releaseNotes,
                            "binaryType" to release.binaryType.name
                        ))
                    }
                }.addOnFailureListener {
                    result.error("CHECK_FAILED", "Can not check for new release", "checkForNewRelease() failed with $it")
                }
            }
            "updateApp" -> {
                firebaseAppDistribution.updateApp()
                    .addOnProgressListener { progress ->
                        emitProgress(progress.apkBytesDownloaded, progress.apkFileTotalBytes, progress.updateStatus.name)
                    }
                    .addOnSuccessListener {
                        result.success(null)
                    }
                    .addOnFailureListener {
                        result.error("UPDATE_FAILED", "Can not update app", "updateApp() failed with $it")
                    }
            }
            "isNewReleaseAvailable" -> {
                firebaseAppDistribution.checkForNewRelease().addOnSuccessListener { release ->
                    result.success(release != null)
                }.addOnFailureListener {
                    result.error("CHECK_FAILED", "Can not check for new release", "checkForNewRelease() failed with $it")
                }
            }
            "isTesterSignedIn" -> {
                result.success(firebaseAppDistribution.isTesterSignedIn)
            }
            "signInTester" -> {
                firebaseAppDistribution.signInTester().addOnSuccessListener {
                    result.success(true)
                }.addOnFailureListener {
                    result.error("SIGN_IN_TESTER_FAILED", "Can not sign in tester", "signInTester() failed with $it")
                }
            }
            "signOutTester" -> {
                firebaseAppDistribution.signOutTester()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        progressChannel.setStreamHandler(null)
        eventSink = null
        context = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun emitProgress(apkBytesDownloaded: Long, apkFileTotalBytes: Long, updateStatus: String) {
        eventSink?.success(mapOf(
            "apkBytesDownloaded" to apkBytesDownloaded,
            "apkFileTotalBytes" to apkFileTotalBytes,
            "updateStatus" to updateStatus
        ))
    }
}

File: firebase_app_distribution_ios/ios/firebase_app_distribution_ios/Sources/firebase_app_distribution_ios/FirebaseAppDistributionPlugin.swift
import Flutter
import UIKit
import FirebaseAppDistribution

public class FirebaseAppDistributionPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var latestDownloadURL: URL?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "firebase_app_distribution_ios", binaryMessenger: registrar.messenger())
    let progressChannel = FlutterEventChannel(name: "firebase_app_distribution_ios/download_progress", binaryMessenger: registrar.messenger())
    let instance = FirebaseAppDistributionPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    progressChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      switch call.method {
        case "updateIfNewReleaseAvailable":
          AppDistribution.appDistribution().checkForUpdate(completion: { release, error in
            if error != nil {
                // Handle error
                result(FlutterError(code: "CHECK_FAILED", message: "Can not check for new release", details: error?.localizedDescription))
                return
            }

            guard let release = release else {
              result(nil)
              return
            }

            self.latestDownloadURL = release.downloadURL
            let title = "New Version Available"
            let message = "Version \(release.displayVersion)(\(release.buildVersion)) is available."
            let uialert = UIAlertController(title: title,message: message, preferredStyle: .alert)

            uialert.addAction(UIAlertAction(title: "Update", style: UIAlertAction.Style.default) {
              _ in
              UIApplication.shared.open(release.downloadURL)
            })
            uialert.addAction(UIAlertAction(title: "Cancel", style: UIAlertAction.Style.cancel) {
              _ in
            })

            if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
                rootViewController.present(uialert, animated: true, completion: nil)
            }
            result(nil)
          })

      case "checkForNewRelease":
          AppDistribution.appDistribution().checkForUpdate(completion: { release, error in
              if error != nil {
                  result(FlutterError(code: "CHECK_FAILED", message: "Can not check for new release", details: error?.localizedDescription))
                  return
              }

              guard let release = release else {
                  self.latestDownloadURL = nil
                  result(nil)
                  return
              }

              self.latestDownloadURL = release.downloadURL
              result([
                  "displayVersion": release.displayVersion,
                  "buildVersion": release.buildVersion,
                  "downloadUrl": release.downloadURL.absoluteString,
              ])
          })

      case "updateApp":
          guard let downloadURL = latestDownloadURL else {
              result(FlutterError(code: "UPDATE_NOT_AVAILABLE", message: "No checked release is available to install", details: nil))
              return
          }

          UIApplication.shared.open(downloadURL)
          result(nil)

      case "isNewReleaseAvailable":
          if (!AppDistribution.appDistribution().isTesterSignedIn) {
              result(false)
              return
          }

          AppDistribution.appDistribution().checkForUpdate(completion: { release, error in
              if error != nil {
                  // Handle error
                  result(FlutterError(code: "CHECK_FAILED", message: "Can not check for new release", details: error?.localizedDescription))
                  return
              }

              guard let release = release else {
                result(false)
                return
              }

              result(true)
          })

      case "isTesterSignedIn":
          result(AppDistribution.appDistribution().isTesterSignedIn)

      case "signInTester":
          AppDistribution.appDistribution().signInTester(completion: {error in
              if error != nil {
                  // Handle error
                  result(FlutterError(code: "SIGN_IN_TESTER_FAILED", message: "Can not sign in tester", details: error?.localizedDescription))
                  return
              }

              result(nil)
          })

      case "signOutTester":
          AppDistribution.appDistribution().signOutTester()
          result(nil)

      default:
          result(nil)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    return nil
  }
}
